code/test/checkUtil.js

```diff
@@ -0,0 +1,45 @@
+// get OneWallet state
+const getONEWalletState = async (wallet) => {
```

Comment:
If we want to compare the state against something (especially javascript object) we might want to parse each of the state variables instead of using object returned by web3/truffle. But for informational purposes this looks fine for now

Reply:
for an example see [here](https://github.com/polymorpher/one-wallet/blob/master/code/lib/api/index.js#L262)

Reply:
@polymorpher Thanks for the example :pray:
Have updated `getONEWalletState` and `checkONEWallet` to use plain JS objects in the [latest commit](https://github.com/1wallet-dev/1wallet/commit/d6a4e8e11dcd7f82beba8ff54a97c68882986463)

@polymorpher could you review specifically the following conversions for nonce and lastOperationTime
```
  // const nonce = new BN(await wallet.getNonce())
  const nonce = (await wallet.getNonce()).toNumber()
  const lastOperationTime = new BN(await wallet.lastOperationTime())
  ```
  The goal is to be able to easily compare the state values agains those returned by the function call and update the state to the new values where possible.

Reply:
Updating and will push shortly
Currently for all numbers we're converting them to BN and then toNumber()
e.g.
`const nonce = new BN(await wallet.getNonce()).toNumber()`
except for SpendingStates which are stored in wei and if we use this approach we get an exception parsing
`Failed to parse walletSpendingState: Error: Number can only safely store up to 53 bits`
as a workaround we are converting these back to ETH first using
`spendingLimit: new BN(unit.fromWei(walletSpendingState[0], 'ether')).toNumber(),`
`spendingAmount: new BN(unit.fromWei(walletSpendingState[1], 'ether')).toNumber(),`
`highestSpendingLimit: new BN(unit.fromWei(walletSpendingState[5], 'ether')).toNumber(),`

below is an example of the state returned
```
{
    "address": "0x2F62d7CF38576709828537D9E7e5DebE1f6E4750",
    "identificationKey": "0xd9a9eb4f14cde1ada8c3b04b063b0da893ae137e1ed8fdd52310908f44bb5740fd38f62e32072fe93cd1e763dd8dd95ed6c8eadc5f49fd5d57c68671266137e0",
    "identificationKeys": {
        "0": {
            "key": "0"
        }
    },
    "forwardAddress": "0x0000000000000000000000000000000000000000",
    "info": {
        "root": "0xe670b093536fdd441c35c9811bd2d29ee441b945bf9582804596803b11e9977e",
        "height": 5,
        "interval": 30,
        "t0": 54921498,
        "lifespan": 12,
        "maxOperationsPerInterval": 1,
        "recoveryAddress": "0x2B0178e0F3fBC9c0f43825FdC14845dcEBb50F3F",
        "extra": 0
    },
    "oldInfo": {},
    "innerCores": {
        "0": {
            "root": "0xe0d92235fc0f9dac939d213214dde318baaa5a1fb89bc48924bfea2ff5fb02c7",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "1": {
            "root": "0x1dbeeb8c2d1ef9f31db9deebac0ace41d16a4c13bbeddd0d09e51e25b862fc12",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "2": {
            "root": "0x4d2faff76619f7679f56f9abfd12b020faddbb1d193584db8710bb106a22f717",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "3": {
            "root": "0x7ea9df9143ecdaf59348fa4fab1c50ef88e9a56b6bdeb39ae3885c7521524ade",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "4": {
            "root": "0xc41607660b9658c4448b537619cefc1d0117237a13c44bdee64a06548ad7e13f",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        "5": {
            "root": "0xab7a676d7bd2de67100b05dd1bc478b431dbe52f905ded264e60e4a93fe7be40",
            "height": 2,
            "interval": 180,
            "t0": 9153583,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        }
    },
    "rootKey": "0xe670b093536fdd441c35c9811bd2d29ee441b945bf9582804596803b11e9977e",
    "version": {
        "majorVersion": 15,
        "minorVersion": 1
    },
    "spendingState": {
        "spendingLimit": 1,
        "spendingAmount": 0,
        "lastSpendingInterval": 0,
        "spendingInterval": 86400,
        "lastLimitAdjustmentTime": 0,
        "highestSpendingLimit": 1
    },
    "nonce": 1,
    "lastOperationTime": 1647645262,
    "allCommits": {
        "commitHashArray": [
            "0x1f9676488ca85db123df4f1c6e6556ee8a8bf8a8febfd1282535f0299bab634d"
        ],
        "paramHashArray": [
            "0x3ded2f8f86e120dd54f35729ce6841bba4e16cb218bbdec30964a3db53c7c2a2"
        ],
        "veriFicationHashArray": [
            "0x3cccb9b5b60f85486f2d656cefe145d302b25910ccfc8180cea91b3c5e55e12e"
        ],
        "timestampArray": [
            "6235124d"
        ],
        "completedArray": [
            true
        ]
    },
    "trackedTokens": {
        "tokenTypeArray": [
            "0"
        ],
        "contractAddressArray": [
            "0xCb36071971a20714d982f75f44ea95E13BB87e60"
        ],
        "tokenIdArray": [
            "0"
        ]
    },
    "backlinks": {}
}
```

Reply:
Also have changed the `assetTransfer` logic to store the `oldState` and update the `currentState`
and changed `checkONEWallet` to be readOnly simply comparing the `oldState` vs `newState` rather than retrieving and updating the currentState.

This role of testing is to 
1. Identify differences (checking new values are correct when they are deterministic)
2. Update the oldState for these values to the currentState values
3. check that there are no other differences in State (i.e. updated oldState = currentState)

Below are two approaches for doing this 
Approach A - use oldState and CurrentState exclusively.
This puts a heavier reliance that the values retrieved from `getONEWalletState` are accurate
example
```
    // Alice Items that have changed - nonce, spendingState, lastOperationTime, Commits
    assert.notEqual(alice.currentState.nonce, alice.oldState.nonce, 'alice wallet.nonce should have been changed')
    assert.equal(alice.currentState.nonce, alice.oldState.nonce + 1, 'alice wallet.nonce should have been changed')
    alice.oldState.nonce = alice.currentState.nonce
    // spendingState
    assert.equal(alice.currentState.spendingState.spentAmount, (ONE_CENT / 2).toString(), 'alice wallet.spentAmount should have been changed')
    alice.oldState.spendingState.spentAmount = alice.currentState.spendingState.spentAmount
    assert.notEqual(alice.currentState.spendingState.lastSpendingInterval, '0', 'alice wallet.spentAmount should have been changed')
    alice.oldState.spendingState.lastSpendingInterval = alice.currentState.spendingState.lastSpendingInterval
    // lastOperationTime
    assert.notStrictEqual(alice.currentState.lastOperationTime, alice.oldState.lastOperationTime, 'alice wallet.lastOperationTime should have been updated')
    alice.oldState.lastOperationTime = alice.currentState.lastOperationTime
    // commits
    assert.notDeepEqual(alice.currentState.allCommits, alice.oldState.allCommits, 'alice wallet.allCommits should have been updated')
    alice.oldState.allCommits = alice.currentState.allCommits
    // check alice
    await CheckUtil.checkONEWallet(alice)
```
Approach B - retrieve the values again in the test suite and use them wherever possible
This places an additional validation layer on top of `getONEWalletState` but can add complexity to the writing of the harnesses. **Also note: there is not a good solution for complex objects like `allCommits`**
example
```
   // Alice Items that have changed - nonce, spendingState, lastOperationTime, Commits
    const nonce = await await alice.wallet.getNonce()
    assert.notEqual(nonce, alice.oldState.nonce, 'alice wallet.nonce should have been changed')
    assert.equal(nonce.toNumber(), alice.oldState.nonce + 1, 'alice wallet.nonce should have been changed')
    alice.oldState.nonce = nonce.toNumber()
    // spendingState
    const spendingState = await alice.wallet.getSpendingState()
    assert.equal(spendingState.spentAmount, (ONE_CENT / 2).toString(), 'alice wallet.spentAmount should have been changed')
    alice.oldState.spendingState.spentAmount = spendingState.spentAmount
    assert.notEqual(spendingState.lastSpendingInterval, '0', 'alice wallet.spentAmount should have been changed')
    alice.oldState.spendingState.lastSpendingInterval = spendingState.lastSpendingInterval
    // lastOperationTime
    const lastOperationTime = await alice.wallet.lastOperationTime()
    assert.notStrictEqual(lastOperationTime, alice.oldState.lastOperationTime, 'alice wallet.lastOperationTime should have been updated')
    alice.oldState.lastOperationTime = lastOperationTime.toNumber()
    // commits
    const allCommits = await alice.wallet.getAllCommits()
    assert.notDeepEqual(allCommits, alice.oldState.allCommits, 'alice wallet.allCommits should have been updated')
    alice.oldState.allCommits = alice.currentState.allCommits
```

Reply:
anything declared as arrays in IONEWallet should be arrays not objects
e.g. identificationKeys

Reply:
Updated to arrays where appropriate latest example of state below
```
{
    "address": "0x219764E1eDF3F59952880EC979d6292f285A2Eed",
    "identificationKey": "0x0b4df4159e9bb0a059a6ddd8d9e3e8991c4607ce58e98e06fdcbbfc06ea3c92687840b86f07041d37169b05c35c4cdfff117be320e87170fb5834d6c8cdc729c",
    "identificationKeys": [
        "0"
    ],
    "forwardAddress": "0x0000000000000000000000000000000000000000",
    "info": {
        "root": "0x8196b1243c545ca4f1660a2865da8a67dc0b02c34629a7eba84a2d5849d57835",
        "height": 5,
        "interval": 30,
        "t0": 54924540,
        "lifespan": 12,
        "maxOperationsPerInterval": 1,
        "recoveryAddress": "0xC6A240e846683cA22F1200e9bc22ef396aA983c0",
        "extra": 0
    },
    "oldInfo": [],
    "innerCores": [
        {
            "root": "0x5d3a7b7308317ad954cebcf7841f6def609c5a3c6638b2d6f263d0deafeb6220",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0x6f3abce8d5e9471d0160769afefab26869989f006ac7da94c234b539236cf104",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0x9f408eadc7a15456243716c6b031260c677eb4337f2f27e1c5027c88cac1db2f",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0xd0fb7aec8c16167246f7bd65945094666b09231f080c2a8af7a0960be27d0eed",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0x7182904842fc242162431dedb170859b590e46abb3e1e9ce2bdee1f06ac07fdd",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        },
        {
            "root": "0xb8b19a862db699f2832af5a62a61cade4caf7f4e2817b1b4915158a0c3701546",
            "height": 2,
            "interval": 180,
            "t0": 9154090,
            "lifespan": 2,
            "maxOperationsPerInterval": 1
        }
    ],
    "rootKey": "0x8196b1243c545ca4f1660a2865da8a67dc0b02c34629a7eba84a2d5849d57835",
    "version": {
        "majorVersion": 15,
        "minorVersion": 1
    },
    "spendingState": {
        "spendingLimit": "1000000000000000000",
        "spentAmount": "5000000000000000",
        "lastSpendingInterval": "19071",
        "spendingInterval": "86400",
        "lastLimitAdjustmentTime": "0",
        "highestSpendingLimit": "1000000000000000000"
    },
    "nonce": 1,
    "lastOperationTime": 1647736528,
    "allCommits": {
        "commitHashArray": [
            "0x7be9bee01d72f71343f19c1c1ff39a9bd133f8d0704c891a1141184b09164c1d"
        ],
        "paramHashArray": [
            "0x59c3caaeb509f2c7ea7e6b343ea6bc9f855163f185a9279e659fb9c0078811de"
        ],
        "veriFicationHashArray": [
            "0x215feaf23e1ce1c2759fa59842d467eb3760ae10f2b041e2cccf35b038b9e871"
        ],
        "timestampArray": [
            "623676d0"
        ],
        "completedArray": [
            true
        ]
    },
    "trackedTokens": {
        "tokenTypeArray": [
            "0"
        ],
        "contractAddressArray": [
            "0x69bF30aC4B8dE7bD0E3C6C56aF45184082D803F7"
        ],
        "tokenIdArray": [
            "0"
        ]
    },
    "backlinks": []
}
```

Reply:
Okay, some details can be changed later to be consistent with what's stored on the contract and expected at the client